app/scripts/components/auto-save.js

```diff
@@ -0,0 +1,95 @@
+'use strict';
+import React from 'react';
+import { connect } from 'react-redux';
+import * as autosave from '../util/auto-save';
+import { compressChanges } from '../util/compress-changes';
+import { fastForward, updateLocalStore } from '../actions';
+
+export const AutoSave = React.createClass({
+
+  componentWillMount: function () {
+    const unsaved = autosave.getLocalActions();
+    if (unsaved) {
+      this.props.dispatch(updateLocalStore(unsaved));
+    }
+  },
+
+  componentWillReceiveProps: function (newProps) {
+    const { historyId, success } = newProps.save;
+    const { past } = newProps.selection;
+    if (success && autosave.getLocalActions()) {
+      // on save success, if there's any stored items, remove them.
+      this.forget();
+    } else if (past.length && historyId !== past[past.length - 1].historyId) {
+      // if we have unsaved changes, persist them to localStorage.
+      this.store(past, historyId);
+    }
+  },
+
+  store: function (past, historyId) {
+    // compress the past selection array as a single action.
+    const compressed = historyId ? compressChanges(past, historyId) : compressChanges(past);
+    autosave.saveLocalActions(compressed);
+  },
+
+  restore: function () {
+    const { cached } = this.props.save;
+    this.props.dispatch(fastForward(cached));
+    this.forget();
+  },
+
+  forget: function () {
+    this.props.dispatch(updateLocalStore(null));
+    autosave.destroyLocalActions();
+  },
+
+  describeAction: function (action) {
+    if (action.undo && action.redo) return 'Modified';
+    else if (action.undo) return 'Deleted';
+    else return 'Created';
+  },
+
+  renderCached: function (cached) {
+    const items = cached.map(action => (
+      <li key={action.id}><strong>{this.describeAction(action)}</strong>: {action.id}</li>
+    ));
+    return (
+      <div>
+        <ul className='cached'>{items}</ul>
+        <button onClick={this.restore}>Restore</button>
+        <button onClick={this.forget}>Forget about it</button>
+      </div>
+    );
+  },
+
+  render: function () {
+    const { cached } = this.props.save;
+    return (
+      <div className='autosave'>
+        { cached ? <div className='modal__cover'></div> : null }
```

Comment:
Yeah it's the single element violation thing. Also I've used this sort of structure to toggle the modal cover and just apply a class to the actual modal, so it animates into view.